### ISSUE TITLE
Aliyun OSS backend support

### ISSUE THREAD

**Issue by is** [restic/restic#1354]
Alibaba Cloud Object Storage Service([aliyun-oss](https://www.alibabacloud.com/product/oss)) is a cloud object storage similar to Swift.

I am willing to contribute the backend for aliyun-oss. I had finshed a basic workable prototype.

**Comment by is** [comment]
prunc does not work well now.

```
debug enabled
password is correct
counting files in repo
building new index for repo
[0:00]   0 / 0 packs
repository contains 0 packs (0 blobs) with 0B
processed 0 blobs: 0 duplicate blobs, 0B duplicate
load all snapshots
find data that is still in use for 3 snapshots
[0:01] 100.00%  3 / 3 snapshots
number of used blobs is larger than number of available blobs!
Please report this error (along with the output of the 'prune' run) at
https://github.com/restic/restic/issues/new
```

**Comment by is** [comment]
Alibaba cloud (aliyun) is one of largest cloud service provider in China. I don't know why you can't open the product page directly :(. Can you try open the homepage https://www.alibabacloud.com, then find products / Storage&CDN / Object Storage Service. The document link is https://www.alibabacloud.com/help/product/31815.htm?spm=a3c0i.7950270.213127.48.4a21e1deWSZAhY

The patch is not completed now, I will further improve the code, supply more test cases.

**Comment by is** [comment]
I have resolved prune problem. It's cause by a mistake in oss.List function.

@fd0 , Can you help me look up the gofmt error in Travis CI build process.
I'm sure that the location.go is formatted before commit.
I can't get any point by the information from CI log.

**Comment by is** [comment]
This PR is basically done. I added the relevant test code. 
The actual use in these days did not found a new problem.

Alibaba cloud support 5G space and 5G traffic per month OSS free quota for each registered user. [OSS Price](https://www.alibabacloud.com/product/oss#pricing).

@fd0, Can you help me to review the PR. If you still can not open the link, please tell me,
I will submit an issue to Alibaba cloud supporter.

**Comment by fd0** [comment]
Alibaba won't let me register an account without providing a mobile phone number and credit card payment information, and I've decided I won't give that to a Chinese company. Are there any other options to get an account for integration testing? Is someone willing to sponsor such an account?

**Comment by fd0** [comment]
Just for the record: I'm not against adding this backend, we just need to figure out the details with the test account and get the code into shape.

**Comment by hsiangtw** [comment]
I will try to create a test account for you. How can I give you?

**Comment by fd0** [comment]
Let's wait if rclone manages to add support for this type of backend, it's tracked here: https://github.com/ncw/rclone/pull/1920 but is blocked by Aliyun not merging and/or commenting on a PR by @ncw here: https://github.com/aliyun/aliyun-oss-go-sdk/pull/76

We're working on an integration of `rclone` into restic (see #1561), so that we don't have to duplicate the work already done by `rclone` and just use it :)

It's currently not possible with the client library to set the `http.Client`, this is also a blocker for restic, e.g. bandwidth limiting won't work.

**Comment by ray-g** [comment]
A workaround to backup to AliYun OSS is to use the [Minio Backend](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#minio-server) and [Minio OSS Gateway](https://github.com/minio/minio/blob/master/docs/gateway/oss.md) together.

**Comment by luolibin** [comment]
@fd0 I was able to event the basic functions of rclone-oss last year, but I did not pass the integration test.
Can we discuss it?

**Comment by luolibin** [comment]
It seems that Alibaba Cloud does not want to open `http.Client`,look this[aliyun/aliyun-oss-go-sdk#76](https://github.com/aliyun/aliyun-oss-go-sdk/issues/72)

**Comment by fd0** [comment]
Looks like it, yes. That worries me a bit, I can't come up with a reason not to do it properly :disappointed:

**Comment by ncw** [comment]
There has been some progress on the alibabacloud issue, so I hope I can merge the backend soon! https://github.com/aliyun/aliyun-oss-go-sdk/pull/76#issuecomment-443642044

**Comment by fd0** [comment]
Awesome, thanks for your persistence :)

**Comment by ncw** [comment]
It looks like this was fixed in: https://github.com/aliyun/aliyun-oss-go-sdk/commit/34e27f16 - I'll warm up the aliyun backend!

**Comment by ncw** [comment]
Here is the first beta of rclone with aliyun / alibaba cloud oss support. No docs yet! The backend is called "oss" if anyone wants to try restic + rclone + oss:

https://beta.rclone.org/branch/v1.45-094-g3d1958f4-oss-beta/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've discovered after a long time and a lot of code I didn't need to write that Alibaba Cloud (Aliyun) OSS is supported just fine by the Amazon S3 SDK.  So I'm going to abandon the OSS backend above!

Restic seems to use the minio S3 SDK and I didn't immediately find the equivalent parameter.

The magic tweak needed is for the Amazon SDK is  `WithS3ForcePathStyle(false)`

https://github.com/ncw/rclone/blob/781142a73fbf8130b02330d1ba46867bff4996fe/backend/s3/s3.go#L824

So I think with an extra config parameter restic should be able to use oss directly.

That said, rclone will support it very shortly :-)

**Comment by MichaelEischer** [comment]
@ncw https://github.com/restic/restic/pull/2733 updates the minio-go library. This also adds a fix related to aliyun https://github.com/minio/minio-go/pull/1253/files . Can you test whether that adds Aliyun support?

**Comment by ncw** [comment]
I think that might work, yes - it automatically sets the hosting style for Alyiun endpoints so it would solve https://github.com/restic/restic/issues/2528 and https://github.com/restic/restic/pull/2535

**Comment by luolibin** [comment]
I'm sorry, I haven't followed this email in recent months, and haven't worked temporarily. I will try if I have time.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Nick Craig-Wood"<notifications@github.com&gt;;
发送时间:&nbsp;2020年5月25日(星期一) 下午5:29
收件人:&nbsp;"restic/restic"<restic@noreply.github.com&gt;;
抄送:&nbsp;"手写的从前"<75738947@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [restic/restic] Aliyun OSS backend support (#1354)





 
I think that might work, yes - it automatically sets the hosting style for Alyiun endpoints so it would solve #2528 and #2535

**Comment by MichaelEischer** [comment]
@luolibin The updated minio library is included in restic 0.10.0. Could you test whether it works for you?

**Comment by MichaelEischer** [comment]
Closing this as #2535 has also been merged.
